examples/amm/src/lib.rs

```diff
@@ -0,0 +1,165 @@
+use std::convert::Infallible;
+
+use async_graphql::{scalar, Request, Response};
+use fungible::AccountOwner;
+use linera_sdk::base::{Amount, ArithmeticError, ContractAbi, ServiceAbi};
+use linera_views::views::ViewError;
+use matching_engine::Parameters;
+use num_bigint::BigUint;
+use num_traits::{cast::FromPrimitive, ToPrimitive, Zero};
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+
+pub struct AmmAbi;
+
+impl ContractAbi for AmmAbi {
+    type InitializationArgument = ();
+    type Parameters = Parameters;
+    type Operation = Operation;
+    type ApplicationCall = ApplicationCall;
+    type Message = Message;
+    type SessionCall = ();
+    type Response = ();
+    type SessionState = ();
+}
+
+impl ServiceAbi for AmmAbi {
+    type Query = Request;
+    type QueryResponse = Response;
+    type Parameters = Parameters;
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub enum OperationType {
+    Swap {
+        owner: AccountOwner,
+        input_token_idx: u32,
+        input_amount: Amount,
+        output_token_idx: u32,
+    },
+    AddLiquidity {
+        owner: AccountOwner,
+        token0_amount: Amount,
+        token1_amount: Amount,
+    },
+    RemoveLiquidity {
+        owner: AccountOwner,
+        input_token_idx: u32,
+        input_amount: Amount,
+        other_token_idx: u32,
+    },
+}
+
+scalar!(OperationType);
+
+/// Operations that can be sent to the application.
+#[derive(Debug, Serialize, Deserialize)]
+pub enum Operation {
+    ExecuteOperation { operation: OperationType },
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+pub enum Message {
+    ExecuteOperation { operation: OperationType },
+}
```

Comment:
We should distinguish between the owner(s) of the AMM, who are allowed to add or remove liquidity, and the other users, who can only make swaps. (But I guess that will be part of adding bids.)

Reply:
I just talked about this with @ma2bd:
We could also allow anyone to add liquidity, and then keep track of what percentage of the liquidity was added by whom. You'd then be allowed to remove liquidity, but only up to the percentage you had originally added.

Reply:
Talked on Slack, posting resolution here. We'll do three steps:

1. First just enforce that the main chain owner always has 100% of the shares, i.e. only they can add/remove liquidity.
2. As a later improvement, allow anyone to add liquidity and keep track of the percentage (or "shares" or however we represent it) that they contributed.
3. And maybe even later, make the shares into fungible tokens themselves.

This PR will cover point 1